File: libs/location_sharing/location_sharing_session.cpp
#include "location_sharing_session.hpp"

#include "base/logging.hpp"

#include <chrono>

namespace location_sharing
{

// Forward declare from location_sharing_types.cpp
uint64_t GetCurrentTimestamp();

LocationSharingSession::LocationSharingSession() = default;

LocationSharingSession::~LocationSharingSession()
{
  if (IsActive())
    Stop();
}

SessionCredentials LocationSharingSession::Start(SessionConfig const & config)
{
  if (m_state != SessionState::Inactive)
  {
    LOG(LWARNING, ("Session already active, stopping previous session"));
    Stop();
  }

  SetState(SessionState::Starting);

  m_config = config;
  m_credentials = SessionCredentials::Generate();
  m_currentPayload = std::make_unique<LocationPayload>();
  m_lastUpdateTimestamp = 0;

  LOG(LINFO, ("Location sharing session started, ID:", m_credentials.sessionId));

  SetState(SessionState::Active);

  return m_credentials;
}

void LocationSharingSession::Stop()
{
  if (m_state == SessionState::Inactive)
    return;

  SetState(SessionState::Stopping);

  LOG(LINFO, ("Location sharing session stopped"));

  m_currentPayload.reset();
  m_credentials = SessionCredentials();
  m_lastUpdateTimestamp = 0;

  SetState(SessionState::Inactive);
}

void LocationSharingSession::UpdateLocation(location::GpsInfo const & gpsInfo)
{
  if (!IsActive())
  {
    LOG(LWARNING, ("Cannot update location - session not active"));
    return;
  }

  if (!m_currentPayload)
  {
    m_currentPayload = std::make_unique<LocationPayload>();
  }

  // Update location data
  m_currentPayload->timestamp = GetCurrentTimestamp();
  m_currentPayload->latitude = gpsInfo.m_latitude;
  m_currentPayload->longitude = gpsInfo.m_longitude;
  m_currentPayload->accuracy = gpsInfo.m_horizontalAccuracy;

  if (gpsInfo.m_speed > 0.0)
    m_currentPayload->speed = gpsInfo.m_speed;
  else
    m_currentPayload->speed = std::nullopt;

  if (gpsInfo.m_bearing >= 0.0)
    m_currentPayload->bearing = gpsInfo.m_bearing;
  else
    m_currentPayload->bearing = std::nullopt;

  ProcessLocationUpdate();
}

void LocationSharingSession::UpdateNavigationInfo(
    uint64_t eta,
    uint32_t distanceRemaining,
    std::string const & destinationName)
{
  if (!IsActive() || !m_currentPayload)
    return;

  m_currentPayload->mode = SharingMode::Navigation;
  m_currentPayload->eta = eta;
  m_currentPayload->distanceRemaining = distanceRemaining;

  if (m_config.includeDestinationName && !destinationName.empty())
    m_currentPayload->destinationName = destinationName;
}

void LocationSharingSession::ClearNavigationInfo()
{
  if (!m_currentPayload)
    return;

  m_currentPayload->mode = SharingMode::Standalone;
  m_currentPayload->eta = std::nullopt;
  m_currentPayload->distanceRemaining = std::nullopt;
  m_currentPayload->destinationName = std::nullopt;
}

void LocationSharingSession::UpdateBatteryLevel(uint8_t batteryPercent)
{
  if (!IsActive() || !m_currentPayload)
    return;

  if (m_config.includeBatteryLevel)
    m_currentPayload->batteryLevel = batteryPercent;

  // Check if battery is too low
  if (batteryPercent < m_config.lowBatteryThreshold)
  {
    LOG(LINFO, ("Battery level too low (", static_cast<int>(batteryPercent), "%), stopping location sharing"));
    OnError("Battery level too low");
    Stop();
  }
}

void LocationSharingSession::SetState(SessionState newState)
{
  if (m_state == newState)
    return;

  SessionState oldState = m_state;
  m_state = newState;

  LOG(LINFO, ("Location sharing state changed:", static_cast<int>(oldState), "->", static_cast<int>(newState)));

  if (m_stateChangeCallback)
    m_stateChangeCallback(newState);
}

void LocationSharingSession::OnError(std::string const & error)
{
  LOG(LERROR, ("Location sharing error:", error));

  if (m_errorCallback)
    m_errorCallback(error);
}

void LocationSharingSession::ProcessLocationUpdate()
{
  if (!ShouldSendUpdate())
    return;

  auto encryptedPayload = CreateEncryptedPayload();
  if (!encryptedPayload.has_value())
  {
    OnError("Failed to create encrypted payload");
    return;
  }

  m_lastUpdateTimestamp = GetCurrentTimestamp();

  if (m_payloadReadyCallback)
    m_payloadReadyCallback(*encryptedPayload);
}

bool LocationSharingSession::ShouldSendUpdate() const
{
  if (!m_currentPayload)
    return false;

  uint64_t now = GetCurrentTimestamp();
  uint64_t timeSinceLastUpdate = now - m_lastUpdateTimestamp;

  return timeSinceLastUpdate >= m_config.updateIntervalSeconds;
}

std::optional<EncryptedPayload> LocationSharingSession::CreateEncryptedPayload() const
{
  if (!m_currentPayload)
  {
    LOG(LERROR, ("No payload to encrypt"));
    return std::nullopt;
  }

  std::string json = m_currentPayload->ToJson();

  auto encrypted = crypto::EncryptAes256Gcm(m_credentials.encryptionKey, json);
  if (!encrypted.has_value())
  {
    LOG(LERROR, ("Encryption failed"));
    return std::nullopt;
  }

  return encrypted;
}

}  // namespace location_sharing

File: iphone/Maps/Core/LocationSharing/LocationSharingSession.swift
import Foundation
import CoreLocation

/// Session state for location sharing
@objc enum LocationSharingState: Int {
  case inactive
  case starting
  case active
  case paused
  case stopping
  case error
}

/// Mode for location sharing
enum LocationSharingMode {
  case standalone  // GPS only
  case navigation  // GPS + ETA + distance
}

/// Configuration for location sharing session
struct LocationSharingConfig {
  var updateIntervalSeconds: Int = 20
  var includeDestinationName: Bool = true
  var includeBatteryLevel: Bool = true
  var lowBatteryThreshold: Int = 10
  var serverBaseUrl: String = "https://live.organicmaps.app"
}

/// Session credentials
struct LocationSharingCredentials {
  let sessionId: String
  let encryptionKey: String

  /// Generate share URL from credentials
  func generateShareUrl(serverBaseUrl: String) -> String {
    let combined = "\(sessionId):\(encryptionKey)"
    guard let data = combined.data(using: .utf8) else { return "" }
    let base64 = data.base64EncodedString()
      .replacingOccurrences(of: "+", with: "-")
      .replacingOccurrences(of: "/", with: "_")
      .replacingOccurrences(of: "=", with: "")

    var url = serverBaseUrl
    if !url.hasSuffix("/") {
      url += "/"
    }
    url += "live/\(base64)"

    return url
  }

  /// Generate new random credentials
  static func generate() -> LocationSharingCredentials {
    let sessionId = UUID().uuidString
    let encryptionKey = Self.generateRandomKey()
    return LocationSharingCredentials(sessionId: sessionId, encryptionKey: encryptionKey)
  }

  private static func generateRandomKey() -> String {
    var bytes = [UInt8](repeating: 0, count: 32)
    _ = SecRandomCopyBytes(kSecRandomDefault, bytes.count, &bytes)
    return Data(bytes).base64EncodedString()
  }
}

/// Location payload structure
struct LocationPayload {
  var timestamp: TimeInterval
  var latitude: Double
  var longitude: Double
  var accuracy: Double
  var speed: Double?
  var bearing: Double?
  var mode: LocationSharingMode = .standalone
  var eta: TimeInterval?
  var distanceRemaining: Int?
  var destinationName: String?
  var batteryLevel: Int?

  init(location: CLLocation) {
    self.timestamp = Date().timeIntervalSince1970
    self.latitude = location.coordinate.latitude
    self.longitude = location.coordinate.longitude
    self.accuracy = location.horizontalAccuracy

    if location.speed >= 0 {
      self.speed = location.speed
    }

    if location.course >= 0 {
      self.bearing = location.course
    }
  }

  func toJSON() -> [String: Any] {
    var json: [String: Any] = [
      "timestamp": Int(timestamp),
      "lat": latitude,
      "lon": longitude,
      "accuracy": accuracy
    ]

    if let speed = speed {
      json["speed"] = speed
    }

    if let bearing = bearing {
      json["bearing"] = bearing
    }

    json["mode"] = mode == .navigation ? "navigation" : "standalone"

    if mode == .navigation {
      if let eta = eta {
        json["eta"] = Int(eta)
      }
      if let distanceRemaining = distanceRemaining {
        json["distanceRemaining"] = distanceRemaining
      }
      if let destinationName = destinationName {
        json["destinationName"] = destinationName
      }
    }

    if let batteryLevel = batteryLevel {
      json["batteryLevel"] = batteryLevel
    }

    return json
  }
}

/// Main location sharing session manager
@objc class LocationSharingSession: NSObject {

  // Singleton instance
  @objc static let shared = LocationSharingSession()

  // State
  private(set) var state: LocationSharingState = .inactive
  private(set) var credentials: LocationSharingCredentials?
  private(set) var config: LocationSharingConfig = LocationSharingConfig()
  private(set) var shareUrl: String?

  // Current payload
  private var currentPayload: LocationPayload?
  private var lastUpdateTimestamp: TimeInterval = 0

  // Callbacks
  var onStateChange: ((LocationSharingState) -> Void)?
  var onError: ((String) -> Void)?
  var onPayloadReady: ((Data) -> Void)?

  private override init() {
    super.init()
  }

  /// Start location sharing session
  @objc func start(with config: LocationSharingConfig) -> String? {
    if state != .inactive {
      NSLog("Location sharing already active, stopping previous session")
      stop()
    }

    setState(.starting)

    self.config = config
    self.credentials = LocationSharingCredentials.generate()
    self.lastUpdateTimestamp = 0

    guard let credentials = self.credentials else {
      onError?("Failed to generate credentials")
      setState(.error)
      return nil
    }

    self.shareUrl = credentials.generateShareUrl(serverBaseUrl: config.serverBaseUrl)

    NSLog("Location sharing session started: \(credentials.sessionId)")
    setState(.active)

    return shareUrl
  }

  /// Stop location sharing session
  @objc func stop() {
    if state == .inactive {
      return
    }

    setState(.stopping)

    NSLog("Location sharing session stopped")

    currentPayload = nil
    credentials = nil
    shareUrl = nil
    lastUpdateTimestamp = 0

    setState(.inactive)
  }

  /// Update location
  func updateLocation(_ location: CLLocation) {
    guard state == .active else {
      NSLog("Cannot update location - session not active")
      return
    }

    currentPayload = LocationPayload(location: location)
    processLocationUpdate()
  }

  /// Update navigation info
  func updateNavigationInfo(eta: TimeInterval, distanceRemaining: Int, destinationName: String?) {
    guard state == .active, currentPayload != nil else { return }

    currentPayload?.mode = .navigation
    currentPayload?.eta = eta
    currentPayload?.distanceRemaining = distanceRemaining

    if config.includeDestinationName, let name = destinationName {
      currentPayload?.destinationName = name
    }
  }

  /// Clear navigation info
  func clearNavigationInfo() {
    currentPayload?.mode = .standalone
    currentPayload?.eta = nil
    currentPayload?.distanceRemaining = nil
    currentPayload?.destinationName = nil
  }

  /// Update battery level
  func updateBatteryLevel(_ level: Int) {
    guard state == .active else { return }

    if config.includeBatteryLevel {
      currentPayload?.batteryLevel = level
    }

    // Stop if battery too low
    if level < config.lowBatteryThreshold {
      NSLog("Battery level too low (\(level)%), stopping location sharing")
      onError?("Battery level too low")
      stop()
    }
  }

  // MARK: - Private methods

  private func setState(_ newState: LocationSharingState) {
    if state == newState {
      return
    }

    NSLog("Location sharing state: \(state.rawValue) -> \(newState.rawValue)")
    state = newState
    onStateChange?(newState)
  }

  private func processLocationUpdate() {
    guard shouldSendUpdate() else { return }

    guard let encryptedData = createEncryptedPayload() else {
      onError?("Failed to create encrypted payload")
      return
    }

    lastUpdateTimestamp = Date().timeIntervalSince1970
    onPayloadReady?(encryptedData)
  }

  private func shouldSendUpdate() -> Bool {
    guard currentPayload != nil else { return false }

    let now = Date().timeIntervalSince1970
    let timeSinceLastUpdate = now - lastUpdateTimestamp

    return timeSinceLastUpdate >= Double(config.updateIntervalSeconds)
  }

  private func createEncryptedPayload() -> Data? {
    guard let payload = currentPayload,
          let credentials = credentials else {
      return nil
    }

    let json = payload.toJSON()

    guard let jsonData = try? JSONSerialization.data(withJSONObject: json),
          let jsonString = String(data: jsonData, encoding: .utf8) else {
      NSLog("Failed to serialize payload to JSON")
      return nil
    }

    // Call native encryption (via bridge)
    guard let encryptedJson = LocationSharingBridge.encryptPayload(
      key: credentials.encryptionKey,
      plaintext: jsonString) else {
      NSLog("Encryption failed")
      return nil
    }

    return encryptedJson.data(using: .utf8)
  }
}

/// Swift wrapper for LocationSharingBridgeObjC
extension LocationSharingBridge {
  static func encryptPayload(key: String, plaintext: String) -> String? {
    return LocationSharingBridgeObjC.encryptPayload(withKey: key, plaintext: plaintext)
  }
}
